Cesium3DTiles/include/Cesium3DTiles/Cesium3DTiles.h

```diff
@@ -0,0 +1,199 @@
+  static inline constexpr const char* TypeName = "BoundingVolume";
+
+  std::optional<std::vector<double>> box;
+
+  std::optional<std::vector<double>> region;
+
+  std::optional<std::vector<double>> sphere;
+};
+
+struct CODEGEN_API TileContent : public CesiumUtility::ExtensibleObject {
+  static inline constexpr const char* TypeName = "TileContent";
+
+  std::optional<BoundingVolume> boundingVolume;
+
+  std::string uri;
+};
+
+struct CODEGEN_API Tile : public CesiumUtility::ExtensibleObject {
+  static inline constexpr const char* TypeName = "Tile";
+
+  enum class Refine { ADD, REPLACE };
+
+  BoundingVolume boundingVolume;
+
+  std::optional<BoundingVolume> viewerRequestVolume;
+
+  double geometricError;
+
+  std::optional<Refine> refine;
+
+  std::vector<double> transform =
+      {1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1};
+
+  std::optional<TileContent> content;
+
+  std::optional<std::vector<Tile>> children;
+};
+
+struct CODEGEN_API TilesetProperties : public CesiumUtility::ExtensibleObject {
+  static inline constexpr const char* TypeName = "TilesetProperties";
+
+  double maximum;
+
+  double minimum;
+};
+
+struct CODEGEN_API Tileset : public CesiumUtility::ExtensibleObject {
+  static inline constexpr const char* TypeName = "Tileset";
+
+  struct CODEGEN_API Properties : public CesiumUtility::ExtensibleObject {
+    static inline constexpr const char* TypeName = "Properties";
+
+    std::unordered_map<std::string, TilesetProperties> additionalProperties;
+  };
+
+  Asset asset;
+
+  std::optional<Properties> properties;
+
+  double geometricError;
+
+  Tile root;
+
+  std::optional<std::vector<std::string>> extensionsUsed;
+
+  std::optional<std::vector<std::string>> extensionsRequired;
+};
+
+struct CODEGEN_API FeatureTable : public CesiumUtility::ExtensibleObject {
+  static inline constexpr const char* TypeName = "FeatureTable";
+
+  struct CODEGEN_API BinaryBodyReference
+      : public CesiumUtility::ExtensibleObject {
+    static inline constexpr const char* TypeName = "BinaryBodyReference";
+
+    enum class ComponentType {
+      BYTE,
+      UNSIGNED_BYTE,
+      SHORT,
+      UNSIGNED_SHORT,
+      INT,
+      UNSIGNED_INT,
+      FLOAT,
+      DOUBLE
+    };
+
+    double byteOffset;
+
+    std::optional<ComponentType> componentType;
+  };
+
+  struct CODEGEN_API Property {
+    static inline constexpr const char* TypeName = "Property";
+
+    std::optional<BinaryBodyReference> v0;
+    std::optional<std::vector<double>> v1;
+    std::optional<double> v2;
+  };
+  struct CODEGEN_API GlobalPropertyScalar {
+    static inline constexpr const char* TypeName = "GlobalPropertyScalar";
+
+    struct CODEGEN_API Variant0 : public CesiumUtility::ExtensibleObject {
+      static inline constexpr const char* TypeName = "Variant0";
+
+      double byteOffset;
+    };
+
+    std::optional<Variant0> v0;
+    std::optional<std::vector<double>> v1;
+    std::optional<double> v2;
```

Comment:
This may have the same problem with the bounding volume above. The schema specifies GlobalPropertyScalar can only accept one of the properties above. Should using `std::variant` be a better choice in this case? cc @kring for more opinions

Reply:
Did consider this, my concern was that the reader becomes much more complicated since it has to "read every alternative at the same time", so
- some template/STL gymnastics is required to get each type instantiated as a read destination
- there's no obvious choice when multiple alternatives are valid other than "first in the specified order", which also requires some STL mess to accomplish

Not opposed to doing it that way, just will take a little more work and be a little STL-heavy.

The "multiple `optional`s` approach also covers the semantics of both `oneOf` and `anyOf`, granted that's not relevant for glTF/3DTiles

Reply:
Without necessarily understanding the full implications, std::variant sounds right to me. It more closely matches the schema semantics and will be more memory efficient as well.